File: Documentation/5.4/Samples/python/DocumentExtensions/TimeSeries/RollupAndRetention.py
from datetime import datetime

from ravendb.documents.operations.time_series import (
    RawTimeSeriesPolicy,
    TimeSeriesPolicy,
    TimeSeriesCollectionConfiguration,
    TimeSeriesConfiguration,
    ConfigureTimeSeriesOperation,
)
from ravendb.primitives.time_series import TimeValue

from examples_base import ExampleBase


class RollupAndRetention(ExampleBase):
    def setUp(self):
        super().setUp()

    def test_rollup_and_retention(self):
        with self.embedded_server.get_document_store("RollupAndRetention") as store:
            with store.open_session() as session:
                # region rollup_and_retention_0
                # Policy for the original ("raw") time-series,
                # to keep the data for one week
                one_week = TimeValue.of_days(7)
                raw_retention = RawTimeSeriesPolicy(one_week)

                # Roll-up the data for each day,
                # and keep the results for one year
                one_day = TimeValue.of_days(1)
                one_year = TimeValue.of_years(1)
                daily_rollup = TimeSeriesPolicy("DailyRollupForOneYear", one_day, one_year)

                # Enter the above policies into a
                # time-series collection configuration
                # for the collection 'Sales'
                sales_ts_config = TimeSeriesCollectionConfiguration(policies=[daily_rollup], raw_policy=raw_retention)

                # Enter the configuration for the Sales collection
                # into a time-series configuration for the whole database
                database_ts_config = TimeSeriesConfiguration()
                database_ts_config.collections["Sales"] = sales_ts_config

                # Send the time-series configuration to the server
                store.maintenance.send(ConfigureTimeSeriesOperation(database_ts_config))
                # endregion

                # region rollup_and_retention_1
                # Create local instance of the time-series "rawSales"
                # in the document "sales/1"
                raw_ts = session.time_series_for("sales/1", "rawSales")

                # Create local instance of the rollup time-series - first method:
                daily_rollup_TS = session.time_series_for("sales/1", "rawSales@DailyRollupForOneYear")

                # Create local instance of the rollup time-series - second method:
                # using the rollup policy itself and the raw time-series' name
                rollup_time_series_2 = session.time_series_for("sales/1", daily_rollup.get_time_series_name("rawSales"))

                # Retrieve all the data from both time-series
                raw_data = raw_ts.get(datetime.min, datetime.max)
                rollup_data = daily_rollup_TS.get(datetime.min, datetime.max)
                # endregion

File: Documentation/5.4/Samples/php/DocumentExtensions/TimeSeries/RollupAndRetention.php
<?php

namespace RavenDB\Samples\DocumentExtensions\TimeSeries;

use PHPUnit\Framework\TestCase;
use RavenDB\Documents\DocumentStore;
use RavenDB\Documents\Operations\TimeSeries\ConfigureTimeSeriesOperation;
use RavenDB\Documents\Operations\TimeSeries\RawTimeSeriesPolicy;
use RavenDB\Documents\Operations\TimeSeries\TimeSeriesCollectionConfiguration;
use RavenDB\Documents\Operations\TimeSeries\TimeSeriesConfiguration;
use RavenDB\Documents\Operations\TimeSeries\TimeSeriesPolicy;
use RavenDB\Documents\Operations\TimeSeries\TimeSeriesPolicyArray;
use RavenDB\Primitives\TimeValue;

class RollupAndRetention extends TestCase
{
    public function samples(): void
    {
        $store = new DocumentStore("http://localhost:8080", "Examples");
        $store->initialize();

        try {
            $session = $store->openSession();
            try {
                # region rollup_and_retention_0
                $oneWeek = TimeValue::ofDays(7);
                $fiveYears = TimeValue::ofYears(5);

                // Define a policy on the RAW time series data:
                // ============================================
                $rawPolicy = new RawTimeSeriesPolicy($fiveYears); // Retain entries for five years

                // Define a ROLLUP policy:
                // =======================
                $rollupPolicy = new TimeSeriesPolicy(
                        "By1WeekFor1Year", // Name of policy
                        $oneWeek, // Aggregation time, roll-up the data for each week
                        $fiveYears); // Retention time, keep data for five years

                // Define the time series configuration for collection "Companies" (use above policies):
                // =====================================================================================
                $companyConfig = new TimeSeriesCollectionConfiguration();
                $companyConfig->setPolicies(TimeSeriesPolicyArray::fromArray([ $rollupPolicy ]));
                $companyConfig->setRawPolicy($rawPolicy);

                $timeSeriesConfig = new TimeSeriesConfiguration();
                $timeSeriesConfig->setCollections([
                    "Companies" => $companyConfig
                ]);

                // Deploy the time series configuration to the server
                // by sending the 'ConfigureTimeSeriesOperation' operation:
                // ========================================================
                $store->maintenance()->send(new ConfigureTimeSeriesOperation($timeSeriesConfig));

                // NOTE:
                // The time series entries in the RavenDB sample data are dated up to the year 2020.
                // To ensure that you see the rollup time series created when running this example,
                // the retention time should be set to exceed that year.
                # endregion

                # region rollup_and_retention_1
                // Get all data from the RAW time series:
                // ======================================

                $rawData = $session
                    ->timeSeriesFor("companies/91-A", "StockPrices")
                    ->get();

                // Get all data from the ROLLUP time series:
                // =========================================

                // Either - pass the rollup name explicitly to 'TimeSeriesFor':
                $rollupData = $session
                    ->timeSeriesFor("companies/91-A", "StockPrices@By1WeekFor1Year")
                    ->get();

                // Or - get the rollup name by calling 'GetTimeSeriesName':
                $rollupData = $session
                    ->timeSeriesFor("companies/91-A", $rollupPolicy->GetTimeSeriesName("StockPrices"))
                    ->get();

                // The raw time series has 100 entries
                $this->assertCount(100, $rawData);
                $this->assertFalse($rawData[0]->isRollup());

                // The rollup time series has only 22 entries
                // as each entry aggregates 1 week's data from the raw time series
                $this->assertCount(22, $rollupData);
                $this->assertTrue($rollupData[0]->isRollup());
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}
